draft-ietf-tls-tls13.md

```diff
@@ -2793,47 +2798,107 @@ specified for all records when processing early data records.
 Specifically, decryption failure of any 0-RTT record following
 an accepted "early_data" extension MUST produce a fatal
 "bad_record_mac" alert as per {{record-payload-protection}}.
+Implementations SHOULD determine the security parameters for the
+1-RTT phase of the connection entirely before processing the EncryptedExtensions
+and Finished, using those values solely to determine whether to
+accept or reject 0-RTT data.
 
 [[TODO: How does the client behave if the indication is rejected.]]
 
+##### Processing Order
+
+Clients are permitted to "stream" 0-RTT data until they
+receive the server's Finished, only then sending the "end_of_early_data"
+alert. In order to avoid deadlock, when accepting "early_data",
+servers MUST process the client's Finished and then immediately
+send the ServerHello, rather than waiting for the client's
+"end_of_early_data" alert.
 
-##### Replay Properties
+##### Replay Properties {#replay-time}
 
-As noted in {{zero-rtt-data}}, TLS does not provide any
+As noted in {{zero-rtt-data}}, TLS provides only a limited
 inter-connection mechanism for replay protection for data sent by the
-client in the first flight.  As a special case, implementations where
-the server configuration, is delivered out of band (as has been
-proposed for DTLS-SRTP {{RFC5763}}), MAY use a unique server
-configuration identifier for each connection, thus preventing
-replay. Implementations are responsible for ensuring uniqueness of the
-identifier in this case.
+client in the first flight. 
+
+The "ticket_age" extension sent by the client SHOULD be used by
+servers to limit the time over which the first flight might be
+replayed.  A server can store the time at which it sends a server
+configuration to a client, or encode the time in a ticket.  Then, each
+time it receives an early_data extension, it can check to see if the
+value used by the client matches its expectations.
+
+The "ticket_age" value provided by the client will be shorter than the
+actual time elapsed on the server by a single round trip time.  This
+difference is comprised of the delay in sending the NewSessionTicket
+message to the client, plus the time taken to send the ClientHello to
+the server.  For this reason, a server SHOULD measure the round trip
+time prior to sending the NewSessionTicket message and account for
+that in the value it saves.
+
+There are several potential sources of error that make an exact
+measurement of time difficult.  Variations in client and server clocks
+are likely to be minimal, outside of gross time corrections.  Network
+propagation delays are most likely causes of a mismatch in legitimate
+values for elapsed time.  Both the NewSessionTicket and ClientHello
+messages might be retransmitted and therefore delayed, which might be
+hidden by TCP.
```

Comment:
The client could also have since moved to another network. Or perhaps got load-balanced to a different server. Or maybe I got the ticket at 4am when no one was awake and now it's noon and the network's busier.

I'm not sure how useful it is to pin down all the sources of error and recommend that servers SHOULD measure the RTT and put it in NewSessionTicket. I think more realistic is that servers will pick some vaguely reasonable amount of slop and use that.

I also suspect the difference between "an attacker can replay this for 5 seconds" and "an attacker can replay this for 1 minute" isn't much and so keeping the clock tight isn't that valuable. If you're online-ish, you can get a lot of replays in anyway. I'd probably buy "an attacker can replay this for 1 minute" vs. "an attacker can replay this for however long this ticket key is alive" is meaningful since that's now offline-ish.